File: e0509_gripper_description/scripts/gripper_joint_publisher.py
#!/usr/bin/env python3
"""
RViz 시각화용 통합 조인트 상태 발행 노드

역할:
    - dynamic_joint_states에서 로봇 팔 조인트 상태 수신
    - gripper/stroke 토픽에서 그리퍼 stroke 값 수신
    - 로봇 팔 + 그리퍼 조인트를 합쳐서 joint_states로 발행
    - RViz에서 전체 로봇 시각화

Note:
    그리퍼 제어는 gripper_service_node.py가 담당합니다.
    이 노드는 시각화만 담당합니다.
"""
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from control_msgs.msg import DynamicJointState
from std_msgs.msg import Int32


class GripperJointPublisher(Node):
    def __init__(self):
        super().__init__('gripper_joint_publisher')

        # Publisher for joint states
        self.publisher = self.create_publisher(JointState, 'joint_states', 10)
        self.timer = self.create_timer(0.02, self.publish_joint_states)  # 50Hz

        # Robot arm joint names
        self.arm_joint_names = [
            'joint_1', 'joint_2', 'joint_3',
            'joint_4', 'joint_5', 'joint_6'
        ]

        # Gripper joint names
        self.gripper_joint_names = [
            'gripper_rh_r1',
            'gripper_rh_r2',
            'gripper_rh_l1',
            'gripper_rh_l2'
        ]

        # All joint names
        self.all_joint_names = self.arm_joint_names + self.gripper_joint_names

        # Robot arm joint positions (from dynamic_joint_states)
        self.arm_positions = {name: 0.0 for name in self.arm_joint_names}
        self.arm_velocities = {name: 0.0 for name in self.arm_joint_names}

        # Gripper position control
        # stroke: 0 = open, 700 = fully closed (real gripper)
        # joint angle: 0.0 = open, ~1.0 rad = closed
        self.stroke = 0  # Current stroke value (0~700)
        self.target_stroke = 0
        self.stroke_speed = 50  # stroke units per cycle (faster for visualization)

        # Stroke to joint angle conversion
        # stroke 700 → ~1.0 rad
        self.stroke_to_rad = 1.0 / 700.0

        # Dynamic joint states 구독 (로봇 팔 조인트)
        self.dynamic_sub = self.create_subscription(
            DynamicJointState, 'dynamic_joint_states',
            self.dynamic_joint_state_callback, 10)

        # Stroke 토픽 구독 (gripper_service_node가 발행)
        self.stroke_sub = self.create_subscription(
            Int32, 'gripper/stroke', self.stroke_callback, 10)

        self.get_logger().info('========================================')
        self.get_logger().info('Combined Joint Publisher Ready!')
        self.get_logger().info('----------------------------------------')
        self.get_logger().info('Subscribing to:')
        self.get_logger().info('  dynamic_joint_states - Arm joints')
        self.get_logger().info('  gripper/stroke - Int32 (0~700)')
        self.get_logger().info('Publishing to:')
        self.get_logger().info('  joint_states - All joint angles')
        self.get_logger().info('========================================')

    def dynamic_joint_state_callback(self, msg: DynamicJointState):
        """로봇 팔 조인트 상태 수신"""
        for i, name in enumerate(msg.joint_names):
            if name in self.arm_positions and i < len(msg.interface_values):
                interface_val = msg.interface_values[i]
                # position은 보통 첫 번째 인터페이스
                if len(interface_val.values) > 0:
                    self.arm_positions[name] = interface_val.values[0]
                if len(interface_val.values) > 1:
                    self.arm_velocities[name] = interface_val.values[1]

    def stroke_callback(self, msg):
        """토픽으로 stroke 값 수신"""
        stroke = max(0, min(700, msg.data))
        self.target_stroke = stroke
        self.get_logger().debug(f'Received stroke: {stroke}')

    def publish_joint_states(self):
        # Smooth movement towards target for gripper
        if abs(self.stroke - self.target_stroke) > 1:
            if self.stroke < self.target_stroke:
                self.stroke = min(self.stroke + self.stroke_speed, self.target_stroke)
            else:
                self.stroke = max(self.stroke - self.stroke_speed, self.target_stroke)
        else:
            self.stroke = self.target_stroke

        # Convert stroke to joint angle
        gripper_angle = self.stroke * self.stroke_to_rad

        # Build combined joint state message
        msg = JointState()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.name = self.all_joint_names

        # Arm positions
        positions = [self.arm_positions[name] for name in self.arm_joint_names]
        velocities = [self.arm_velocities[name] for name in self.arm_joint_names]

        # Gripper positions (all 4 joints same angle)
        positions.extend([gripper_angle] * 4)
        velocities.extend([0.0] * 4)

        msg.position = positions
        msg.velocity = velocities
        msg.effort = [0.0] * len(self.all_joint_names)
        self.publisher.publish(msg)


def main():
    rclpy.init()
    node = GripperJointPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: e0509_gripper_description/src/gripper_joint_publisher.cpp
/*
 * gripper_joint_publisher.cpp
 * Combined joint state publisher for RViz visualization
 *
 * Combines robot arm joints (from dynamic_joint_states) and gripper joints
 * (from gripper/stroke topic) into a single joint_states message at 50Hz.
 */

#include <algorithm>
#include <cmath>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joint_state.hpp"
#include "control_msgs/msg/dynamic_joint_state.hpp"
#include "std_msgs/msg/int32.hpp"

using namespace std::chrono_literals;

class GripperJointPublisher : public rclcpp::Node {
public:
    GripperJointPublisher()
        : Node("gripper_joint_publisher"),
          stroke_(0.0), target_stroke_(0.0), stroke_speed_(50.0),
          stroke_to_rad_(1.0 / 700.0)
    {
        publisher_ = this->create_publisher<sensor_msgs::msg::JointState>("joint_states", 10);
        timer_ = this->create_wall_timer(20ms, std::bind(&GripperJointPublisher::publish_joint_states, this));

        // Initialize arm joint positions
        for (const auto& name : arm_joint_names_) {
            arm_positions_[name] = 0.0;
            arm_velocities_[name] = 0.0;
        }

        // Subscribe to dynamic_joint_states (arm joints)
        dynamic_sub_ = this->create_subscription<control_msgs::msg::DynamicJointState>(
            "dynamic_joint_states", 10,
            std::bind(&GripperJointPublisher::dynamic_joint_state_callback, this, std::placeholders::_1));

        // Subscribe to gripper/stroke
        stroke_sub_ = this->create_subscription<std_msgs::msg::Int32>(
            "gripper/stroke", 10,
            std::bind(&GripperJointPublisher::stroke_callback, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "========================================");
        RCLCPP_INFO(this->get_logger(), "Combined Joint Publisher Ready!");
        RCLCPP_INFO(this->get_logger(), "----------------------------------------");
        RCLCPP_INFO(this->get_logger(), "Subscribing to:");
        RCLCPP_INFO(this->get_logger(), "  dynamic_joint_states - Arm joints");
        RCLCPP_INFO(this->get_logger(), "  gripper/stroke - Int32 (0~700)");
        RCLCPP_INFO(this->get_logger(), "Publishing to:");
        RCLCPP_INFO(this->get_logger(), "  joint_states - All joint angles");
        RCLCPP_INFO(this->get_logger(), "========================================");
    }

private:
    void dynamic_joint_state_callback(const control_msgs::msg::DynamicJointState::SharedPtr msg) {
        for (size_t i = 0; i < msg->joint_names.size(); ++i) {
            const auto& name = msg->joint_names[i];
            if (arm_positions_.count(name) && i < msg->interface_values.size()) {
                const auto& iface = msg->interface_values[i];
                if (!iface.values.empty()) {
                    arm_positions_[name] = iface.values[0];
                }
                if (iface.values.size() > 1) {
                    arm_velocities_[name] = iface.values[1];
                }
            }
        }
    }

    void stroke_callback(const std_msgs::msg::Int32::SharedPtr msg) {
        target_stroke_ = std::clamp(static_cast<double>(msg->data), 0.0, 700.0);
    }

    void publish_joint_states() {
        // Smooth gripper movement
        if (std::abs(stroke_ - target_stroke_) > 1.0) {
            if (stroke_ < target_stroke_) {
                stroke_ = std::min(stroke_ + stroke_speed_, target_stroke_);
            } else {
                stroke_ = std::max(stroke_ - stroke_speed_, target_stroke_);
            }
        } else {
            stroke_ = target_stroke_;
        }

        double gripper_angle = stroke_ * stroke_to_rad_;

        auto msg = sensor_msgs::msg::JointState();
        msg.header.stamp = this->get_clock()->now();

        // Build joint names
        msg.name.insert(msg.name.end(), arm_joint_names_.begin(), arm_joint_names_.end());
        msg.name.insert(msg.name.end(), gripper_joint_names_.begin(), gripper_joint_names_.end());

        // Arm positions
        for (const auto& name : arm_joint_names_) {
            msg.position.push_back(arm_positions_[name]);
            msg.velocity.push_back(arm_velocities_[name]);
        }

        // Gripper positions (all 4 joints same angle)
        for (size_t i = 0; i < gripper_joint_names_.size(); ++i) {
            msg.position.push_back(gripper_angle);
            msg.velocity.push_back(0.0);
        }

        msg.effort.resize(msg.name.size(), 0.0);
        publisher_->publish(msg);
    }

    const std::vector<std::string> arm_joint_names_ = {
        "joint_1", "joint_2", "joint_3", "joint_4", "joint_5", "joint_6"
    };
    const std::vector<std::string> gripper_joint_names_ = {
        "gripper_rh_r1", "gripper_rh_r2", "gripper_rh_l1", "gripper_rh_l2"
    };

    std::map<std::string, double> arm_positions_;
    std::map<std::string, double> arm_velocities_;
    double stroke_;
    double target_stroke_;
    double stroke_speed_;
    double stroke_to_rad_;

    rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Subscription<control_msgs::msg::DynamicJointState>::SharedPtr dynamic_sub_;
    rclcpp::Subscription<std_msgs::msg::Int32>::SharedPtr stroke_sub_;
};


int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<GripperJointPublisher>();
    rclcpp::spin(node);
    node.reset();
    rclcpp::shutdown();
    return 0;
}
